### PR TITLE
Prune <script> and <form> tags, as well as HTML comments, also when allo...

### PR DESCRIPTION
**Pull request by amatriain**
...wed_attributes or allowed_tags is set.

Fixes https://github.com/rails/rails-html-sanitizer/issues/27

### CODE REVIEW

File: lib/rails/html/sanitizer.rb [hunk 1]
```diff
@@ -112,6 +112,7 @@ def sanitize(html, options = {})
           # No duck typing, Loofah ensures subclass of Loofah::Scrubber
           loofah_fragment.scrub!(scrubber)
         elsif allowed_tags(options) || allowed_attributes(options)
+          remove_xpaths(loofah_fragment, XPATHS_TO_REMOVE)
```
Comment (kaspth) [inline lib/rails/html/sanitizer.rb]: Followup to my comment on #27. I'd be happier if we're conditionally removing the xpaths, like this:
```ruby
@permit_scrubber.tags = allowed_tags(options)
remove_xpaths(loofah_fragment, XPATHS_TO_REMOVE) if @permit_scrubber.tags
```

**Comment by kaspth** [comment]
I would also like a test for this behavior. I.e. set only `allowed_attributes` and the default xpaths are removed.

**Comment by rafaelfranca** [comment]
In fact I'd like to see that if only `allowed_attributes` is set, we would accept only the `allowed_tags`, that in this case should be empty array.

**Comment by kaspth** [comment]
@rafaelfranca that's the current behavior, if I'm reading you right.

**Comment by rafaelfranca** [comment]
If so, why it is allowing `form`?

**Comment by kaspth** [comment]
`form` is an acceptable tag to Loofah: https://github.com/flavorjones/loofah/blob/master/lib/loofah/html5/whitelist.rb#L53

**Comment by amatriain** [comment]
Refined the behavior of WhiteListSanitizer to put it in line with your comments. Reworked tests a bit and added new tests for the changed behavior.

I'm not too happy with test names, too wordy. Suggestions are welcome.

WhiteListSanitizer now works like this:

- if neither allowed_attributes nor allowed_tags are set, <script> and <form> tags are pruned. This is a sane default.
- if allowed_attributes is set and allowed_tags is not set, same as above (prune <script>, <form> tags)
- if allowed_tags is set, any tags in the array are not touched by the sanitizer and other tags are stripped (including <script> and <form> tags)

In all of the above combinations, HTML comments are pruned. The API currently has no option to enable or disable the pruning of comments, and I think it's safer to prune them in all cases by default.

**Comment by amatriain** [comment]
A question related to this PR: why does PermitScrubber strip tags (keeping their contents) instead of pruning them (removing their contents)? This causes a bit of inconsistent behavior: 

- if neither allowed_tags nor allowed_attributes are set, <script> and <form> tags are pruned.
- if allowed_tags is set but e.g. "script" is not in the allowed array, <script> tags are stripped, not pruned

I'd like to address this, changing PermitScrubber so that it prunes unallowed tags. This should be enough to make the sanitizer behave consistently in the above cases. I'm thinking it may be better to address this in a different PR, what do you think?
